In [32]:
import nltk
import nltk.corpus
from nltk.corpus import stopwords
#nltk.download("stopwords")
from nltk.tokenize import word_tokenize
from nltk.corpus import PlaintextCorpusReader

import random
random.seed(42)

from sklearn.pipeline import Pipeline
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfTransformer

from sklearn.naive_bayes import MultinomialNB
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.svm import SVC

from sklearn.metrics import classification_report, confusion_matrix


import pandas as pd
import numpy as np
import string

#pd.set_option("display.max_rows",500)

import warnings
warnings.filterwarnings('ignore')

In [2]:
stop_words = set(stopwords.words('english'))

In [3]:
corpus = pd.read_csv('data/corpus.csv')
corpus.head()

Unnamed: 0                  Party  \
0  George Washington           Unaffiliated   
1         John Adams             Federalist   
2   Thomas Jefferson  Democratic-Republican   
3      James Madison  Democratic-Republican   
4       James Monroe  Democratic-Republican   

                                         transcripts  
0  Fellow Citizens of the Senate and the House of...  
1  When it was first perceived, in early times, t...  
2  FRIENDS AND FELLOW-CITIZENS, Called upon to un...  
3  Unwilling to depart from examples of the most ...  
4  I should be destitute of feeling if I was not ...

In [33]:
speeches = pd.read_csv('data/presidential_speeches.csv')
speeches.head()

Date          President         Party  \
0  1789-04-30  George Washington  Unaffiliated   
1  1789-10-03  George Washington  Unaffiliated   
2  1790-01-08  George Washington  Unaffiliated   
3  1790-12-08  George Washington  Unaffiliated   
4  1790-12-29  George Washington  Unaffiliated   

                                        Speech Title  \
0                            First Inaugural Address   
1                          Thanksgiving Proclamation   
2                   First Annual Message to Congress   
3                  Second Annual Message to Congress   
4  Talk to the Chiefs and Counselors of the Senec...   

                                             Summary  \
0  Washington calls on Congress to avoid local an...   
1  At the request of Congress, Washington establi...   
2  In a wide ranging speech, President Washington...   
3  Washington focuses on commerce in his second a...   
4  The President reassures the Seneca Nation that...   

                                          Transcript  \
0  Fellow Citizens of the Senate and the House of...   
1  Whereas it is the duty of all Nations to ackno...   
2  Fellow Citizens of the Senate and House of Rep...   
3  Fellow citizens of the Senate and House of Rep...   
4  I the President of the United States, by my ow...   

                                                 URL  
0  https://millercenter.org/the-presidency/presid...  
1  https://millercenter.org/the-presidency/presid...  
2  https://millercenter.org/the-presidency/presid...  
3  https://millercenter.org/the-presidency/presid...  
4  https://millercenter.org/the-presidency/presid...

mention or show example of Rscript to divide text

In [5]:
def document_features(document): 
    document_words = set(document) 
    features = {}
    for word in word_features:
        features['contains({})'.format(word)] = (word in document_words)
    return features

In [6]:
ob=open('data/obama1.txt','rU')
raw_o=ob.read()
raw_o=raw_o.lower()
word_tokens = word_tokenize(raw_o)
obama = [word for word in word_tokens if word not in stop_words]
obama = [word for word in obama if word.isalpha()]

In [7]:
wa=open('data/washington.txt','rU')
raw_wa=wa.read()
raw_wa=raw_wa.lower()
word_tokens = word_tokenize(raw_wa)
washington = [word for word in word_tokens if word not in stop_words]
washington = [word for word in washington if word.isalpha()]

In [8]:
len(obama)/1000

98.401

In [9]:
len(washington)/1000

14.751

In [10]:
obama1=[]
for i in range(98):
    obama1.append([obama[i*1000:(i+1)*1000], 'ob'])

In [11]:
washington1=[]
for i in range(14):
    washington1.append([washington[i*1000:(i+1)*1000], 'wa'])

In [12]:
wo = obama+washington

In [13]:
speeches = washington1+obama1

In [14]:
all_words = nltk.FreqDist(w.lower() for w in wo)
word_features = list(all_words)[:200]

In [15]:
random.shuffle(speeches)

In [16]:
featuresets = [(document_features(d), c) for (d,c) in speeches]
len(featuresets)

112

In [17]:
train_set, test_set = featuresets[50:], featuresets[:50]
classifier = nltk.NaiveBayesClassifier.train(train_set)

In [18]:
print(nltk.classify.accuracy(classifier, test_set))

1.0


roosevelt

In [19]:
ro=open('data/roosevelt.txt','rU')
raw_ro=ro.read()
raw_ro=raw_ro.lower()
word_tokens = word_tokenize(raw_ro)
roosevelt = [word for word in word_tokens if word not in stop_words]
roosevelt = [word for word in roosevelt if word.isalpha()]

In [20]:
len(roosevelt)/1000

63.053

In [21]:
roosevelt1=[]
for i in range(63):
    roosevelt1.append([roosevelt[i*1000:(i+1)*1000], 'ro'])

In [22]:
rwo = obama+washington+roosevelt
speeches = washington1+obama1+roosevelt1

In [23]:
all_words = nltk.FreqDist(w.lower() for w in rwo)
word_features = list(all_words)[:200]

In [24]:
random.shuffle(speeches)

In [25]:
featuresets = [(document_features(d), c) for (d,c) in speeches]
len(featuresets)

175

In [26]:
train_set, test_set = featuresets[50:], featuresets[:50]
classifier = nltk.NaiveBayesClassifier.train(train_set)
print(nltk.classify.accuracy(classifier, test_set))

0.98


Kennedy

In [27]:
ke=open('data/kennedy.txt','rU')
raw_ke=ke.read()
raw_ke=raw_ke.lower()
word_tokens = word_tokenize(raw_ke)
kennedy = [word for word in word_tokens if word not in stop_words]
kennedy = [word for word in kennedy if word.isalpha()]

In [28]:
len(kennedy)/1000

58.108

In [29]:
kennedy1=[]
for i in range(58):
    kennedy1.append([kennedy[i*1000:(i+1)*1000], 'ke'])

In [30]:
rwok = obama+washington+roosevelt+kennedy
speeches = washington1+obama1+roosevelt1+kennedy1

In [31]:
all_words = nltk.FreqDist(w.lower() for w in rwok)
word_features = list(all_words)[:200]

In [32]:
random.shuffle(speeches)
featuresets = [(document_features(d), c) for (d,c) in speeches]
len(featuresets)

233

In [33]:
train_set, test_set = featuresets[50:], featuresets[:50]
classifier = nltk.NaiveBayesClassifier.train(train_set)
print(nltk.classify.accuracy(classifier, test_set))

0.92


In [34]:
#train_set

Clinton

In [35]:
cl=open('data/clinton.txt','rU')
raw_cl=cl.read()
raw_cl=raw_cl.lower()
word_tokens = word_tokenize(raw_cl)
clinton = [word for word in word_tokens if word not in stop_words]
clinton = [word for word in clinton if word.isalpha()]

In [36]:
len(clinton)/1000

66.506

In [37]:
clinton1=[]
for i in range(58):
    clinton1.append([clinton[i*1000:(i+1)*1000], 'ke'])

In [38]:
rwokc = obama+washington+roosevelt+kennedy+clinton
speeches = washington1+obama1+roosevelt1+kennedy1+clinton1

In [39]:
all_words = nltk.FreqDist(w.lower() for w in rwokc)
word_features = list(all_words)[:200]

In [40]:
random.shuffle(speeches)
featuresets = [(document_features(d), c) for (d,c) in speeches]
len(featuresets)

291

In [41]:
train_set, test_set = featuresets[50:], featuresets[:50]
classifier = nltk.NaiveBayesClassifier.train(train_set)
print(nltk.classify.accuracy(classifier, test_set))

0.84


Reagan

In [42]:
re=open('data/reagan.txt','rU')
raw_re=re.read()
raw_re=raw_re.lower()
word_tokens = word_tokenize(raw_re)
reagan = [word for word in word_tokens if word not in stop_words]
reagan = [word for word in reagan if word.isalpha()]

In [43]:
len(reagan)/1000

99.653

In [44]:
reagan1=[]
for i in range(58):
    reagan1.append([reagan[i*1000:(i+1)*1000], 'ke'])

In [45]:
rwokcr = obama+washington+roosevelt+kennedy+clinton+reagan
speeches = washington1+obama1+roosevelt1+kennedy1+clinton1+reagan1

In [46]:
all_words = nltk.FreqDist(w.lower() for w in rwokcr)
word_features = list(all_words)[:200]

In [47]:
random.shuffle(speeches)
featuresets = [(document_features(d), c) for (d,c) in speeches]
len(featuresets)

349

In [48]:
train_set, test_set = featuresets[50:], featuresets[:50]
classifier = nltk.NaiveBayesClassifier.train(train_set)
print(nltk.classify.accuracy(classifier, test_set))

0.82


## Other Classifiers

We will now look at TFIDF score for each speech and use the president label to train other classifiers. We will see if we can improve upon the results above. 

## Logistic Regression

The logistic regression was very helpful in spam/ham classifer, it was again very helpful in comparsion to several other model types. 

In [34]:
speeches = speeches[speeches.President.isin(
    ["George Washington", 'Franklin D. Roosevelt',
     "Barack Obama",'John F. Kennedy',
     'Bill Clinton','Ronald Reagan'])]

In [35]:
def text_process(mess):
    """
    Takes in a string of text, then performs the following:
    1. Remove all punctuation
    2. Remove all stopwords
    3. Returns a list of the cleaned text
    """
    nopunc = [char for char in mess if char not in string.punctuation]
    nopunc = ''.join(nopunc)
    
    return [word for word in nopunc.split() 
            if word.lower() 
            not in stopwords.words('english')]

In [48]:
X_train, X_test, y_train, y_test = \
    train_test_split(speeches.Transcript, 
                     speeches.President, 
                     test_size=0.30,
                     random_state=random.seed(42),
                     stratify=speeches.President)

In [57]:
pipeline = Pipeline([
    ('bow', CountVectorizer(analyzer=text_process)),  
    ('tfidf', TfidfTransformer()),  
    ('classifier', LogisticRegression(multi_class='multinomial',
                                     solver="lbfgs"))  
])

In [58]:
pipeline.fit(X_train,y_train)
predictions = pipeline.predict(X_test)

In [59]:
print(classification_report(predictions,y_test))

                       precision    recall  f1-score   support

         Barack Obama       1.00      0.75      0.86         4
         Bill Clinton       0.50      1.00      0.67         1
Franklin D. Roosevelt       1.00      1.00      1.00         3
    George Washington       1.00      1.00      1.00         1
      John F. Kennedy       1.00      1.00      1.00         2
        Ronald Reagan       1.00      1.00      1.00         3

             accuracy                           0.93        14
            macro avg       0.92      0.96      0.92        14
         weighted avg       0.96      0.93      0.94        14



In [60]:
print(confusion_matrix(predictions,y_test))

[[3 1 0 0 0 0]
 [0 1 0 0 0 0]
 [0 0 3 0 0 0]
 [0 0 0 1 0 0]
 [0 0 0 0 2 0]
 [0 0 0 0 0 3]]


Here we see that a multi-class logistic regression model performs very well with 93% accuracy and 96% recall. The model had a hard time separating Obama and Clinton. 